backend/src/main/java/com/bakdata/conquery/models/jobs/ImportJob.java

```diff
@@ -63,88 +55,237 @@
+
+		for (Column column : table.getColumns()) {
+
+			if (column.getType() != MajorTypeId.STRING) {
+				continue;
 			}
 
-			log.trace("Begin reading data.");
+			// Might not have an underlying Dictionary (eg Singleton, direct-Number)
+			// but could also be an error :/ Most likely the former
+			if (!dicts.containsKey(column.getName()) || dicts.get(column.getName()) == null) {
+				log.trace("No Dictionary for {}", column);
+				continue;
+			}
+
+
+			if (column.getSharedDictionary() != null) {
+				// If the column is based on a shared dict. We reference a new empty dictionary or the existing one
+				// but without updated entries. The entries are updated later on, see ImportJob#applyDictionaryMappings.
+				Dictionary sharedDict = null;
+				synchronized (storage.getLockDummy()) {
+					sharedDict = storage.getDictionary(new DictionaryId(table.getDataset().getId(), column.getSharedDictionary()));
+					if (sharedDict == null) {
+						sharedDict = new MapDictionary(table.getDataset(), column.getSharedDictionary());
+						storage.updateDictionary(sharedDict);
+					}
+				}
+				out.put(new DictionaryId(Dataset.PLACEHOLDER.getId(), dicts.get(column.getName()).getName()), sharedDict);
+				continue;
+			}
+
+			// Its a normal dictionary (only valid for this column in this import)
+
+			final Dictionary dict = dicts.get(column.getName());
+			final String name = computeDefaultDictionaryName(importName, column);
+
+
+			out.put(new DictionaryId(Dataset.PLACEHOLDER.getId(), dict.getName()), dict);
+
 
-			container = parser.readData();
+			dict.setDataset(table.getDataset());
+			dict.setName(name);
 		}
 
-		if (container.isEmpty()) {
-			log.warn("Import was empty. Skipping.");
-			getProgressReporter().done();
-			return;
+
+		return out;
+	}
+
+	/**
+	 * Import all dictionaries. Shared dictionaries are merge into existing ones. All are distributed to corresponding workers.
+	 * Create mappings for shared dictionaries dict.
+	 * This is not synchronized because the methods is called within the job execution.
+	 */
+	private static Map<String, DictionaryMapping> importDictionaries(Namespace namespace, Map<String, Dictionary> dicts, Column[] columns, String importName) {
+
+		// Empty Maps are Coalesced to null by Jackson
+		if (dicts == null) {
+			return Collections.emptyMap();
 		}
 
+		final Map<String, DictionaryMapping> out = new HashMap<>();
+
+		log.trace("Importing Dictionaries");
+
+		for (Column column : columns) {
+
+			if (column.getType() != MajorTypeId.STRING) {
+				continue;
+			}
+
+			// Might not have an underlying Dictionary (eg Singleton, direct-Number)
+			// but could also be an error :/ Most likely the former
+			final Dictionary importDictionary = dicts.get(column.getName());
+			if (!dicts.containsKey(column.getName()) || importDictionary == null) {
+				log.trace("No Dictionary for {}", column);
+				continue;
+			}
+
+
+			if (column.getSharedDictionary() == null) {
```

Comment:
Ich hatte den Code damals eben weil es zuviel auf einmal macht und sich sonst nicht viel teilt bewusst aufgeteilt in normal und shared dictionaries. Der Körper dieser Schleife ist einfach zu kompliziert.

Reply:
Ist unterteilt und ich habe es und die Column Klasse geschoben,